### ISSUE TITLE
Support broadcasting of sample x param

### ISSUE THREAD

**Issue by fritzo** [probtorch/pytorch#79]
@alicanb observed that `Normal` supports broadcasting whereas `Categorical` does not:
```py
p = Variable(torch.Tensor([[0.1, 0.2, 0.3],[0.2, 0.2, 0.3]]), requires_grad=False)
dist = torch.distributions.Categorical(probs=p)
log_prob = dist.log_prob(dist.sample().unsqueeze(-1))
#RuntimeError: The expanded size of the tensor (1) must match the existing size (2) at non-singleton dimension 1
```

**Comment by neerajprad** [comment]
@fritzo, @alicanb - Isn't this issue due to categorical having event dimensions (which are not broadcasted) vs. normal not having any event dimensions which allows for broadcasting along the batch (parameter) dimensions? We started with only allowing for broadcasting along the sample dimensions (leftmost), and normal would have thrown an exception here too. This is also the broadcasting behavior in Tensorflow. With https://github.com/pytorch/pytorch/pull/4210, we decided to make our semantics more flexible and opened up the batch dimensions for broadcasting (primarily to get around scalar dimensionality issues). Does this seem unexpected or confusing?
